docs/rfcs/2023-03-29-file-external-table/rfc.md

```diff
@@ -0,0 +1,86 @@
+---
+Feature Name: "File external table"
+Tracking Issue: https://github.com/GreptimeTeam/greptimedb/issues/1041
+Date: 2023-03-08
+Author: "Xu Wenkang <wenymedia@gmail.com>"
+---
+
+File external table
+
+---
+
+# Summary
+
+Allows users to perform SQL queries on files
+
+# Motivation
+
+User data may already exist in other storages, i.g., file systems/s3, etc. in CSV, parquet, JSON format, etc. We can provide users the ability to perform SQL queries on these files.
+
+
+# Details
+## Overview
+![overview](external-table-engine-overview.png)
+
+We implement a file table engine that creates an external table by accepting user-specified file paths and treating all records as immutable.
+
+1. File Format Decoder: decode files to the `RecordBatch` stream.
+2. Table engine: stores `RecordBatch` in memory and provides scan ability.
+
+## Life cycle
+
+### Register table
+1. Creates metadata and stores it in memory.
+2. Creates the table via file table engine.
+3. Registers table to `CatalogProvider` and it won't register table to `SystemCatalog` (a place that will persist all user-created tables).
+4. Registers table to `SessionLayer` (currently not exists).
+
+### Clean up
+For now, GreptimeDB still lacks a session layer. The system can't release an external table automatically when the user session is closed. But it still has serval ways to allow users to release an external table. 
+1. We enable users to remove an external table manually. 
+2. After the database restarts, all external tables will disappear.
+
+#### Release external tables when starts
+The system  won't persist external tables to disk. Therefore, all external tables wouldn't be recovered after the system restarts.
+
+#### Deregister table
+1. Fetches the target table from `CatalogProvider`
+2. Deregisters the target table in `CatalogProvider`
+3. Finds the target table engine
+4. Drops the target table
+
+## Drawbacks
+- Unsupported filter pushdown
+- Hard to support indexing
+
+# Alternatives
+
+## Using DataFusion API
+We can use datafusion API to register a file table:
+```rust
+let ctx = SessionContext::new();
+
+ctx.register_csv("example", "tests/data/example.csv", CsvReadOptions::new()).await?;
+
+// create a plan
+let df = ctx.sql("SELECT a, MIN(b) FROM example WHERE a <= b GROUP BY a LIMIT 100").await?;
+```
+### Drawbacks
+The DataFusion implements its own `Object Store` abstraction and supports parsing the partitioned directories, which can push down the filter and skips some directories. However, this makes it impossible to use our's `LruCacheLayer`(The parsing of the partitioned directories required paths as input). If we want to manage memory entirely, we should implement our own `TableProvider` or `Table` (We can benefit from reusing the logic of DataFusion's `register_batch` and `MemTable`).
```

Comment:
Is the work too much if we try to implement our own external file `TableProvider` or `Table`? What are the drawbacks of it?

Reply:
We can utilize the `Reader` in the `arrow-csv`/`arrow-csv`/`parquet`(it decodes the target format to `RecordBatch`). For implementing `TableProvider,` we only need to provide the scan ability. That doesn't require too much engineering overhead.